### PR TITLE
[lld][WebAssembly] Report unsupported PIC relocations as errors

### PR DESCRIPTION
**Pull request by luc-blaeser**
`WASM_MEMORY_ADDR_REL_` and `WASM_TABLE_INDEX_REL_` relocations against **undefined symbols** are not supported and, except for `UnresolvedPolicy::ReportError`, lead to incorrect Wasm code, such as invalid data address or invalid table index that cannot be patched during later dynamic Wasm linking with modules declaring those symbols. This is different to other relocations that support undefined symbols by declaring correspond Wasm imports.

For more robust behavior, `wasm-ld` should probably report an error for such unsupported PIC relocations, independent of the `UnresolvedPolicy`.

### CODE REVIEW

File: lld/wasm/Relocations.cpp [hunk 1]
```diff
@@ -170,6 +170,19 @@ void scanRelocations(InputChunk *chunk) {
...
+      case R_WASM_MEMORY_ADDR_REL_SLEB64:
+        // These relocation types are only present in the code section and
+        // are not supported. It would require replacing the constant by using
+        // a GOT global.
+        if (sym->isUndefined())
+          error(toString(file) + ": relocation " +
+                relocTypeToString(reloc.Type) +
+                " is not supported against an undefined symbol `" +
+                toString(*sym) + "`");
+        break;
```
Comment (sbc100) [inline lld/wasm/Relocations.cpp:185]: The above condition looks wrong for this block.   Shouldn't this be just guarded by `sym->isUndefined()` and nothing else?

We don't want to be reporting here is here for defined symbols in PIC mode (for example).    In fact, IIUC, these `REL` relocations are specifically designed for PIC mode.

Comment (luc-blaeser) [reply lld/wasm/Relocations.cpp:188]: Thank you. I adjusted this.

File: lld/test/wasm/unsupported-pic-relocations.s [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+# RUN:   FileCheck %s
+
+# RUN: not wasm-ld --experimental-pic -shared %t.o -o /dev/null  --unresolved-symbols=ignore-all 2>&1 | \
+# RUN:   FileCheck %s
+
+# RUN: not wasm-ld --experimental-pic -shared %t.o -o /dev/null  --unresolved-symbols=import-dynamic 2>&1 | \
+# RUN:   FileCheck %s
+
+.globaltype __memory_base, i32, immutable
+.globaltype	__table_base, i32, immutable
```
Comment (sbc100) [inline lld/test/wasm/unsupported-pic-relocations.s:19]: Indentation looks odd here,  can you make these two lines match.

Comment (sbc100) [reply lld/test/wasm/unsupported-pic-relocations.s:19]: Do you actually need `__memory_base` and `__table_base` at all to make this test work?  Can you use do `i32.const external_func@TBREL` on its own without adding it to `__table_base`?

**Review by sbc100 — COMMENTED** [review]
lgtm % some test cleanups.   I think we can maybe make the tests much smaller.

File: lld/wasm/Relocations.cpp [hunk 3]
```diff
@@ -173,6 +173,22 @@ void scanRelocations(InputChunk *chunk) {
...
 
+    if (sym->isUndefined()) {
+      switch (reloc.Type) {
+      case R_WASM_TABLE_INDEX_REL_SLEB:
+      case R_WASM_TABLE_INDEX_REL_SLEB64:
+      case R_WASM_MEMORY_ADDR_REL_SLEB:
+      case R_WASM_MEMORY_ADDR_REL_SLEB64:
+        // These relocation types are only present in the code section and
+        // are not supported for undefined symbols. It would require replacing
+        // the constant by using a GOT global.
```
Comment (sbc100) [inline lld/wasm/Relocations.cpp:184]: How about:

```
These relocation types are for symbols that exists relative to `__memory_base` or `__table_base` and as such
only make sense for defined symbols.
```

Comment (luc-blaeser) [reply lld/wasm/Relocations.cpp:184]: Thanks. I improved the comment.

**Comment by luc-blaeser** [comment]
Ping

**Comment by dschuff** [comment]
I don't see another revision after you responded to the comments. Did you forget to push it?

**Comment by luc-blaeser** [comment]
> I don't see another revision after you responded to the comments. Did you forget to push it?

Sorry, my mistake, I forgot to push.

**Review by dschuff — APPROVED** [review]

**Comment by dschuff** [comment]
Thanks! Do need me to commit this?

File: lld/test/wasm/unsupported-pic-relocations.s [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+# RUN:   FileCheck %s
+
+# RUN: not wasm-ld --experimental-pic -shared %t.o -o /dev/null  --unresolved-symbols=import-dynamic 2>&1 | \
+# RUN:   FileCheck %s
+
+.functype external_func () -> ()
+
+use_undefined_function:
+    .functype use_undefined_function () -> ()
+    i32.const external_func@TBREL
```
Comment (sbc100) [inline lld/test/wasm/unsupported-pic-relocations.s:22]: I'm still curious how you were able to generate such relocations?  IIUC, LLVM will not generate these relocations for undefined symbols.  Did you come across this issue by writing asm code?  Or is there some issue where LLVM can generate these wrongly?

Comment (luc-blaeser) [reply lld/test/wasm/unsupported-pic-relocations.s:22]: I was using Rust and emitted LLVM IR that I manually compiled down to a Wasm object file.  During testing, I once missed a LLVM IR input file and then saw those relocations on unused functions, without getting an error. It is certainly an edge case.

**Comment by luc-blaeser** [comment]
> Thanks! Do need me to commit this?

Thank you also for your time! Please feel free to merge when you think it is okay. I believe I have no right to merge.

**Review by sbc100 — APPROVED** [review]

**Comment by dschuff** [comment]
Thanks for the contribution!

**Comment by dschuff** [comment]
This seems to have broken the Emscripten benchmark test `test_zzz_bullet`. (Build log [here](https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8736423805323262737/+/u/Emscripten_benchmark/stdout)

The command line is `/b/s/w/ir/x/w/install/bin/wasm-ld -o .libs/libLinearMath.so.0.0.0 .libs/btQuickprof.o .libs/btGeometryUtil.o .libs/btAlignedAllocator.o .libs/btSerializer.o .libs/btConvexHull.o .libs/btConvexHullComputer.o -L/b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten /b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten/libGL-getprocaddr.a -lal -lhtml5 -lstubs-debug -lc-debug /b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten/libdlmalloc.a -lcompiler_rt /b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++-noexcept.a /b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++abi-noexcept.a -lsockets -L/b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten --relocatable -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr`

It looks like this command is trying to build a DSO but I also see it's using `--relocatable` and not `-shared`.
@sbc100 is this what that test is supposed to do? Maybe the test is broken somehow?

There are also warnings like 
`em++: warning: linking a library with `-shared` will emit a static object file.  This is a form of emulation to support existing build systems.  If you want to build a runtime shared library use the SIDE_MODULE setting. [-Wemcc]
em++: warning: ignoring unsupported linker flag: `-soname` [-Wlinkflags]`

which also seems maybe wrong.

**Comment by sbc100** [comment]
Interesting.. all those errors are about `__dso_handle`, which by definition is always available at static link time.  

I think the issue here is that we probably don't want to be reporting these errors in `--relocatable` mode (i.e. when we are just making another object file).

**Comment by sbc100** [comment]
Fix is in https://github.com/llvm/llvm-project/pull/109822

**Pull request merged**